File: src/rider/main/kotlin/com/jetbrains/rider/plugins/specflowriderplugin/psi/GherkinLexer.kt
// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.jetbrains.rider.plugins.specflowriderplugin.psi

import com.intellij.lexer.LexerBase
import com.intellij.openapi.util.text.Strings
import com.intellij.psi.*
import com.intellij.psi.tree.IElementType

class GherkinLexer(private val myKeywordProvider: GherkinKeywordProvider) : LexerBase() {
    private var myBuffer: CharSequence = Strings.EMPTY_CHAR_SEQUENCE
    private var myStartOffset = 0
    private var myEndOffset = 0
    private var myPosition = 0
    private var myCurrentToken: IElementType? = null
    private var myCurrentTokenStart = 0
    private var myKeywords: ArrayList<String>? = null
    private var myState = 0
    private var myCurLanguage: String? = null

    init {
        updateLanguage("en")
    }

    private fun updateLanguage(language: String) {
        myCurLanguage = language
        myKeywords = ArrayList(myKeywordProvider.getAllKeywords(language))
        myKeywords?.sortWith { o1: String, o2: String -> o2.length - o1.length }
    }

    override fun start(buffer: CharSequence, startOffset: Int, endOffset: Int, initialState: Int) {
        myBuffer = buffer
        myStartOffset = startOffset
        myEndOffset = endOffset
        myPosition = startOffset
        myState = initialState
        advance()
    }

    override fun getState(): Int {
        return myState
    }

    override fun getTokenType(): IElementType? {
        return myCurrentToken
    }

    override fun getTokenStart(): Int {
        return myCurrentTokenStart
    }

    override fun getTokenEnd(): Int {
        return myPosition
    }

    private fun isStepParameter(currentElementTerminator: String): Boolean {
        var pos = myPosition
        if (myBuffer[pos] == '<') {
            while (pos < myEndOffset && myBuffer[pos] != '\n' && myBuffer[pos] != '>' && !isStringAtPosition(
                    currentElementTerminator,
                    pos
                )
            ) {
                pos++
            }
            return pos < myEndOffset && myBuffer[pos] == '>'
        }
        return false
    }

    override fun advance() {
        if (myPosition >= myEndOffset) {
            myCurrentToken = null
            return
        }
        myCurrentTokenStart = myPosition
        val c = myBuffer[myPosition]
        if (myState != STATE_INSIDE_PYSTRING && Character.isWhitespace(c)) {
            advanceOverWhitespace()
            myCurrentToken = TokenType.WHITE_SPACE
            while (myPosition < myEndOffset && Character.isWhitespace(myBuffer[myPosition])) {
                advanceOverWhitespace()
            }
        } else if (c == '|' && myState != STATE_INSIDE_PYSTRING) {
            myCurrentToken = GherkinTokenTypes.PIPE
            myPosition++
            myState = STATE_TABLE
        } else if (myState == STATE_PARAMETER_INSIDE_PYSTRING) {
            if (c == '>') {
                myState = STATE_INSIDE_PYSTRING
                myPosition++
                myCurrentToken = GherkinTokenTypes.STEP_PARAMETER_BRACE
            } else {
                advanceToParameterEnd(PYSTRING_MARKER)
                myCurrentToken = GherkinTokenTypes.STEP_PARAMETER_TEXT
            }
        } else if (myState == STATE_INSIDE_PYSTRING) {
            if (isStringAtPosition(PYSTRING_MARKER)) {
                myPosition += 3 /* marker length */
                myCurrentToken = GherkinTokenTypes.PYSTRING
                myState = STATE_DEFAULT
            } else {
                if (myBuffer[myPosition] == '<') {
                    if (isStepParameter(PYSTRING_MARKER)) {
                        myPosition++
                        myState = STATE_PARAMETER_INSIDE_PYSTRING
                        myCurrentToken = GherkinTokenTypes.STEP_PARAMETER_BRACE
                    } else {
                        myPosition++
                        advanceToParameterOrSymbol(PYSTRING_MARKER, STATE_INSIDE_PYSTRING, false)
                        myCurrentToken = GherkinTokenTypes.PYSTRING_TEXT
                    }
                } else {
                    advanceToParameterOrSymbol(PYSTRING_MARKER, STATE_INSIDE_PYSTRING, false)
                    myCurrentToken = GherkinTokenTypes.PYSTRING_TEXT
                }
            }
        } else if (myState == STATE_TABLE) {
            myCurrentToken = GherkinTokenTypes.TABLE_CELL
            while (myPosition < myEndOffset) {
                // Cucumber: 0.7.3 Table cells can now contain escaped bars - \| and escaped backslashes - \\
                if (myBuffer[myPosition] == '\\') {
                    val nextPos = myPosition + 1
                    if (nextPos < myEndOffset) {
                        val nextChar = myBuffer[nextPos]
                        if (nextChar == '|' || nextChar == '\\') {
                            myPosition += 2
                            continue
                        }
                        // else - common case
                    }
                } else if (myBuffer[myPosition] == '|' || myBuffer[myPosition] == '\n') {
                    break
                }
                myPosition++
            }
            while (myPosition > 0 && Character.isWhitespace(myBuffer[myPosition - 1])) {
                myPosition--
            }
        } else if (c == '#') {
            myCurrentToken = GherkinTokenTypes.COMMENT
            advanceToEOL()
            val commentText = myBuffer.subSequence(myCurrentTokenStart + 1, myPosition).toString().trim { it <= ' ' }
            val language = fetchLocationLanguage(commentText)
            language?.let { updateLanguage(it) }
        } else if (c == ':' && myState != STATE_AFTER_STEP_KEYWORD) {
            myCurrentToken = GherkinTokenTypes.COLON
            myPosition++
        } else if (c == '@') {
            myCurrentToken = GherkinTokenTypes.TAG
            myPosition++
            while (myPosition < myEndOffset && isValidTagChar(myBuffer[myPosition])) {
                myPosition++
            }
        } else if (isStringAtPosition(PYSTRING_MARKER)) {
            myCurrentToken = GherkinTokenTypes.PYSTRING
            myState = STATE_INSIDE_PYSTRING
            myPosition += 3
        } else {
            if (myState == STATE_DEFAULT) {
                for (keyword in myKeywords!!) {
                    val length = keyword.length
                    if (isStringAtPosition(keyword)) {
                        if (myKeywordProvider.isSpaceRequiredAfterKeyword(
                                myCurLanguage,
                                keyword
                            ) && myEndOffset - myPosition > length &&
                            Character.isLetterOrDigit(myBuffer[myPosition + length])
                        ) {
                            continue
                        }
                        val followedByChar: Char =
                            if (myPosition + length < myEndOffset) myBuffer[myPosition + length] else 0.toChar()
                        myCurrentToken = myKeywordProvider.getTokenType(myCurLanguage, keyword)
                        if (myCurrentToken === GherkinTokenTypes.STEP_KEYWORD) {
                            val followedByWhitespace = Character.isWhitespace(followedByChar) && followedByChar != '\n'
                            if (followedByWhitespace != myKeywordProvider.isSpaceRequiredAfterKeyword(
                                    myCurLanguage,
                                    keyword
                                )
                            ) {
                                myCurrentToken = GherkinTokenTypes.TEXT
                            }
                        }
                        myPosition += length
                        myState = if (myCurrentToken === GherkinTokenTypes.STEP_KEYWORD) {
                            STATE_AFTER_STEP_KEYWORD
                        } else if (myCurrentToken === GherkinTokenTypes.SCENARIO_OUTLINE_KEYWORD) {
                            STATE_AFTER_SCENARIO_KEYWORD
                        } else {
                            STATE_AFTER_KEYWORD
                        }
                        return
                    }
                }
            }
            if (myState == STATE_PARAMETER_INSIDE_STEP) {
                if (c == '>') {
                    myState = STATE_AFTER_STEP_KEYWORD
                    myPosition++
                    myCurrentToken = GherkinTokenTypes.STEP_PARAMETER_BRACE
                } else {
                    advanceToParameterEnd("\n")
                    myCurrentToken = GherkinTokenTypes.STEP_PARAMETER_TEXT
                }
                return
            } else if (isParameterAllowed) {
                if (myPosition < myEndOffset && myBuffer[myPosition] == '<' && isStepParameter("\n")) {
                    myState = STATE_PARAMETER_INSIDE_STEP
                    myPosition++
                    myCurrentToken = GherkinTokenTypes.STEP_PARAMETER_BRACE
                } else {
                    myCurrentToken = GherkinTokenTypes.TEXT
                    advanceToParameterOrSymbol("\n", STATE_AFTER_STEP_KEYWORD, true)
                }
                return
            }
            myCurrentToken = GherkinTokenTypes.TEXT
            advanceToEOL()
        }
    }

    private val isParameterAllowed: Boolean
        get() = myState == STATE_AFTER_STEP_KEYWORD || myState == STATE_AFTER_SCENARIO_KEYWORD

    private fun advanceOverWhitespace() {
        if (myBuffer[myPosition] == '\n') {
            myState = STATE_DEFAULT
        }
        myPosition++
    }

    private fun isStringAtPosition(keyword: String?): Boolean {
        val length = keyword!!.length
        return myEndOffset - myPosition >= length && myBuffer.subSequence(myPosition, myPosition + length)
            .toString() == keyword
    }

    private fun isStringAtPosition(keyword: String, position: Int): Boolean {
        val length = keyword.length
        return myEndOffset - position >= length && myBuffer.subSequence(position, position + length)
            .toString() == keyword
    }

    private fun advanceToEOL() {
        myPosition++
        val mark = myPosition
        while (myPosition < myEndOffset && myBuffer[myPosition] != '\n') {
            myPosition++
        }
        returnWhitespace(mark)
        myState = STATE_DEFAULT
    }

    private fun returnWhitespace(mark: Int) {
        while (myPosition > mark && Character.isWhitespace(myBuffer[myPosition - 1])) {
            myPosition--
        }
    }

    private fun advanceToParameterOrSymbol(s: String, parameterState: Int, shouldReturnWhitespace: Boolean) {
        val mark = myPosition
        while (myPosition < myEndOffset && !isStringAtPosition(s) && !isStepParameter(s)) {
            myPosition++
        }
        if (shouldReturnWhitespace) {
            myState = STATE_DEFAULT
            if (myPosition < myEndOffset) {
                if (!isStringAtPosition(s)) {
                    myState = parameterState
                }
            }
            returnWhitespace(mark)
        }
    }

    private fun advanceToParameterEnd(endSymbol: String) {
        myPosition++
        val mark = myPosition
        while (myPosition < myEndOffset && !isStringAtPosition(endSymbol) && myBuffer[myPosition] != '>') {
            myPosition++
        }
        if (myPosition < myEndOffset) {
            if (isStringAtPosition(endSymbol)) {
                myState = STATE_DEFAULT
            }
        }
        returnWhitespace(mark)
    }

    override fun getBufferSequence(): CharSequence {
        return myBuffer
    }

    override fun getBufferEnd(): Int {
        return myEndOffset
    }

    companion object {
        private const val STATE_DEFAULT = 0
        private const val STATE_AFTER_KEYWORD = 1
        private const val STATE_TABLE = 2
        private const val STATE_AFTER_STEP_KEYWORD = 3
        private const val STATE_AFTER_SCENARIO_KEYWORD = 4
        private const val STATE_INSIDE_PYSTRING = 5
        private const val STATE_PARAMETER_INSIDE_PYSTRING = 6
        private const val STATE_PARAMETER_INSIDE_STEP = 7
        const val PYSTRING_MARKER = "\"\"\""
        fun fetchLocationLanguage(commentText: String): String? {
            return if (commentText.startsWith("language:")) {
                commentText.substring(9).trim { it <= ' ' }
            } else null
        }

        private fun isValidTagChar(c: Char): Boolean {
            return !Character.isWhitespace(c) && c != '@'
        }
    }
}

File: src/dotnet/ReSharperPlugin.ReqnrollRiderPlugin/Psi/GherkinLexer.cs
using System;
using System.Collections.Generic;
using JetBrains.ReSharper.Psi.Parsing;
using JetBrains.Text;
using JetBrains.Util;
using ReSharperPlugin.ReqnrollRiderPlugin.Caching.ReqnrollJsonSettings;

namespace ReSharperPlugin.ReqnrollRiderPlugin.Psi;

public class GherkinLexer : ILexer
{
    private int _currentPosition;
    private int _myCurrentTokenStart;
    private uint _myState;
    private int _myEndOffset;
    private int _myPystringIdent;
    private int _lastNewLineOffset;
    private string _myCurLanguage;
    private string _closingPyStringMarker;

    private IReadOnlyCollection<string> _myKeywords;
    private readonly GherkinKeywordProvider _keywordProvider;

    // ReSharper disable InconsistentNaming
    private const int STATE_DEFAULT = 0;
    private const int STATE_AFTER_KEYWORD = 1;
    private const int STATE_TABLE = 2;
    private const int STATE_AFTER_KEYWORD_WITH_PARAMETER = 3;
    private const int STATE_INSIDE_PYSTRING = 5;

    private const int STATE_PARAMETER_INSIDE_PYSTRING = 6;
    private const int STATE_PARAMETER_INSIDE_STEP = 7;

    private const string PYSTRING_QUOTE_MARKER = "\"\"\"";
    private const string PYSTRING_BACKTICK_MARKER = "```";
    // ReSharper restore InconsistentNaming

    public GherkinLexer(IBuffer buffer, GherkinKeywordProvider keywordProvider, ReqnrollSettingsProvider reqnrollSettingsProvider)
    {
        Buffer = buffer;
        _keywordProvider = keywordProvider;

        UpdateLanguage(reqnrollSettingsProvider.GetDefaultSettings().Language.NeutralFeature);
    }

    public void Start()
    {
        _myEndOffset = Buffer.Length;
        _currentPosition = 0;
        _myState = 0;
        Advance();
    }

    public void UpdateLanguage(string language)
    {
        if (string.IsNullOrWhiteSpace(language))
            return;
        _myCurLanguage = language;
        _myKeywords = _keywordProvider.GetAllKeywords(language);
    }

    public void Advance()
    {
        if (_currentPosition >= _myEndOffset)
        {
            TokenType = null;
            return;
        }

        _myCurrentTokenStart = _currentPosition;
        char c = Buffer[_currentPosition];
        if (char.IsWhiteSpace(c) && (IsNewLine(out _) || !InPyStringText()))
        {
            TokenType = GherkinTokenTypes.WHITE_SPACE;
            if (AdvanceNewLine())
                return;
            _currentPosition++;

            while (_currentPosition < _myEndOffset && char.IsWhiteSpace(Buffer[_currentPosition]) && !IsNewLine(out _) && !InPyStringText())
                _currentPosition++;
        }
        else if (c == '|' && _myState != STATE_INSIDE_PYSTRING)
        {
            TokenType = GherkinTokenTypes.PIPE;
            _currentPosition++;
            _myState = STATE_TABLE;
        }
        else if (_myState == STATE_PARAMETER_INSIDE_PYSTRING)
        {
            if (c == '>')
            {
                _myState = STATE_INSIDE_PYSTRING;
                _currentPosition++;
                TokenType = GherkinTokenTypes.STEP_PARAMETER_BRACE;
            }
            else
            {
                AdvanceToParameterEnd(_closingPyStringMarker);
                TokenType = GherkinTokenTypes.STEP_PARAMETER_TEXT;
            }
        }
        else if (_myState == STATE_INSIDE_PYSTRING)
        {
            if (IsStringAtPosition(_closingPyStringMarker))
            {
                _currentPosition += 3 /* marker length */;
                TokenType = GherkinTokenTypes.PYSTRING;
                _myState = STATE_DEFAULT;
            }
            else
            {
                if (Buffer[_currentPosition] == '<')
                {
                    if (IsStepParameter(_closingPyStringMarker))
                    {
                        _currentPosition++;
                        _myState = STATE_PARAMETER_INSIDE_PYSTRING;
                        TokenType = GherkinTokenTypes.STEP_PARAMETER_BRACE;
                    }
                    else
                    {
                        _currentPosition++;
                        AdvancePystring();
                        TokenType = GherkinTokenTypes.PYSTRING_TEXT;
                    }
                }
                else
                {
                    AdvancePystring();
                    TokenType = GherkinTokenTypes.PYSTRING_TEXT;
                }
            }
        }
        else if (_myState == STATE_TABLE)
        {
            // We're right after a '|', meaning the start of a Table Cell
            TokenType = GherkinTokenTypes.TABLE_CELL;
                
            // Loop until the end of the cell
            while (_currentPosition < _myEndOffset)
            {
                // Cucumber: 0.7.3 Table cells can now contain escaped bars - \| and escaped backslashes - \\
                if (Buffer[_currentPosition] == '\\')
                {
                    int nextPos = _currentPosition + 1;
                    if (nextPos < _myEndOffset)
                    {
                        char nextChar = Buffer[nextPos];
                        if (nextChar == '|' || nextChar == '\\')
                        {
                            _currentPosition += 2;
                            continue;
                        }

                        // else - common case
                    }
                }
                else if (Buffer[_currentPosition] == '\n')
                {
                    // We reached the end of the line without finding the next '|'
                    // Turns out this wasn't a table cell, just some trailing text
                    // Consider it a comment rather than a table cell, or it will result in an inconsistent cell count error (cf https://github.com/reqnroll/Reqnroll.Rider/issues/12)
                    TokenType = GherkinTokenTypes.COMMENT;
                    break;
                }
                else if (Buffer[_currentPosition] == '|')
                {
                    // End of the cell found
                    break;
                }

                _currentPosition++;
            }

            while (_currentPosition > 0 && char.IsWhiteSpace(Buffer[_currentPosition - 1]))
            {
                _currentPosition--;
            }
        }
        else if (c == '#')
        {
            TokenType = GherkinTokenTypes.COMMENT;
            AdvanceToEol();

            string commentText = Buffer.GetText(new TextRange(_myCurrentTokenStart + 1, _currentPosition)).Trim();
            string language = FetchLocationLanguage(commentText);
            if (language != null)
            {
                UpdateLanguage(language);
            }
        }
        else if (c == ':')
        {
            TokenType = GherkinTokenTypes.COLON;
            _currentPosition++;
        }
        else if (c == '@' && State != STATE_AFTER_KEYWORD && State != STATE_AFTER_KEYWORD_WITH_PARAMETER)
        {
            TokenType = GherkinTokenTypes.TAG;
            _currentPosition++;
            while (_currentPosition < _myEndOffset && IsValidTagChar(Buffer[_currentPosition]))
            {
                _currentPosition++;
            }
        }
        else if (IsStringAtPosition(PYSTRING_QUOTE_MARKER))
        {
            TokenType = GherkinTokenTypes.PYSTRING;
            _myState = STATE_INSIDE_PYSTRING;
            _myPystringIdent = _currentPosition - _lastNewLineOffset;
            _currentPosition += 3;
            _closingPyStringMarker = PYSTRING_QUOTE_MARKER;
        }
        else if (IsStringAtPosition(PYSTRING_BACKTICK_MARKER))
        {
            TokenType = GherkinTokenTypes.PYSTRING;
            _myState = STATE_INSIDE_PYSTRING;
            _myPystringIdent = _currentPosition - _lastNewLineOffset;
            _currentPosition += 3;
            _closingPyStringMarker = PYSTRING_BACKTICK_MARKER;
        }
        else
        {
            if (_myState == STATE_DEFAULT)
            {
                foreach (var keyword in _myKeywords)
                {
                    int length = keyword.Length;
                    if (IsStringAtPosition(keyword))
                    {
                        if (_keywordProvider.IsSpaceRequiredAfterKeyword(_myCurLanguage, keyword) &&
                            _myEndOffset - _currentPosition > length &&
                            char.IsLetterOrDigit(Buffer[_currentPosition + length]))
                        {
                            continue;
                        }

                        char followedByChar = _currentPosition + length < _myEndOffset ? Buffer[_currentPosition + length] : (char)0;
                        TokenType = _keywordProvider.GetTokenType(_myCurLanguage, keyword);
                        if (TokenType == GherkinTokenTypes.STEP_KEYWORD)
                        {
                            bool followedByWhitespace = char.IsWhiteSpace(followedByChar) && followedByChar != '\n';
                            if (followedByWhitespace != _keywordProvider.IsSpaceRequiredAfterKeyword(_myCurLanguage, keyword))
                            {
                                TokenType = GherkinTokenTypes.TEXT;
                            }
                        }

                        _currentPosition += length;
                        if (TokenType == GherkinTokenTypes.STEP_KEYWORD || TokenType == GherkinTokenTypes.SCENARIO_OUTLINE_KEYWORD)
                        {
                            _myState = STATE_AFTER_KEYWORD_WITH_PARAMETER;
                        }
                        else
                        {
                            _myState = STATE_AFTER_KEYWORD;
                        }

                        return;
                    }
                }
            }

            if (_myState == STATE_PARAMETER_INSIDE_STEP)
            {
                if (c == '>')
                {
                    _myState = STATE_AFTER_KEYWORD_WITH_PARAMETER;
                    _currentPosition++;
                    TokenType = GherkinTokenTypes.STEP_PARAMETER_BRACE;
                }
                else
                {
                    AdvanceToParameterEnd("\n");
                    TokenType = GherkinTokenTypes.STEP_PARAMETER_TEXT;
                }

                return;
            }
            else if (_myState == STATE_AFTER_KEYWORD_WITH_PARAMETER)
            {
                if (_currentPosition < _myEndOffset && Buffer[_currentPosition] == '<' && IsStepParameter("\n"))
                {
                    _myState = STATE_PARAMETER_INSIDE_STEP;
                    _currentPosition++;
                    TokenType = GherkinTokenTypes.STEP_PARAMETER_BRACE;
                }
                else
                {
                    TokenType = GherkinTokenTypes.TEXT;
                    AdvanceToParameterOrSymbol("\n", STATE_AFTER_KEYWORD_WITH_PARAMETER, true);
                }

                return;
            }

            TokenType = GherkinTokenTypes.TEXT;
            AdvanceToEol();
        }
    }

    private bool InPyStringText()
    {
        if (_myState != STATE_INSIDE_PYSTRING)
            return false;
        return _currentPosition - _lastNewLineOffset >= _myPystringIdent;
    }

    private bool IsNewLine(out int newLineLength)
    {
        newLineLength = 0;
        if (Buffer[_currentPosition] == '\n')
        {
            newLineLength = 1;
            return true;
        }
        if (Buffer[_currentPosition] == '\r' && _currentPosition + 1 < Buffer.Length && Buffer[_currentPosition + 1] == '\n')
        {
            newLineLength = 2;
            return true;
        }
        return false;
    }

    private bool AdvanceNewLine()
    {
        if (IsNewLine(out var len))
        {
            TokenType = GherkinTokenTypes.NEW_LINE;
            _currentPosition += len;
            _lastNewLineOffset = _currentPosition;
            if (_myState != STATE_INSIDE_PYSTRING)
                _myState = STATE_DEFAULT;
            return true;
        }
        return false;
    }

    private static string FetchLocationLanguage(string commentText)
    {
        return commentText.StartsWith("language:") ? commentText.Substring(9).Trim() : null;
    }

    private bool IsStringAtPosition(string keyword)
    {
        int length = keyword.Length;
        return _myEndOffset - _currentPosition >= length && Buffer.GetText(new TextRange(_currentPosition, _currentPosition + length)).Equals(keyword, StringComparison.Ordinal);
    }

    private bool IsStringAtPosition(string keyword, int position)
    {
        int length = keyword.Length;
        return _myEndOffset - position >= length && Buffer.GetText(new TextRange(position, position + length)).Equals(keyword, StringComparison.Ordinal);
    }

    private static bool IsValidTagChar(char c)
    {
        return !char.IsWhiteSpace(c) && c != '@';
    }

    private void AdvanceToParameterEnd(string endSymbol)
    {
        _currentPosition++;
        int mark = _currentPosition;
        while (_currentPosition < _myEndOffset && !IsStringAtPosition(endSymbol) && Buffer[_currentPosition] != '>')
        {
            _currentPosition++;
        }

        if (_currentPosition < _myEndOffset)
        {
            if (IsStringAtPosition(endSymbol))
            {
                _myState = STATE_DEFAULT;
            }
        }

        ReturnWhitespace(mark);
    }

    private void ReturnWhitespace(int mark)
    {
        while (_currentPosition > mark && char.IsWhiteSpace(Buffer[_currentPosition - 1]))
            _currentPosition--;
    }

    private void AdvanceToEol()
    {
        _currentPosition++;
        int mark = _currentPosition;
        while (_currentPosition < _myEndOffset && Buffer[_currentPosition] != '\n')
        {
            _currentPosition++;
        }
        ReturnWhitespace(mark);
        _myState = STATE_DEFAULT;
    }

    private void AdvancePystring()
    {
        while (_currentPosition < _myEndOffset
               && !IsStepParameter(_closingPyStringMarker)
               && !IsNewLine(out _))
        {
            _currentPosition++;
        }
    }

    private void AdvanceToParameterOrSymbol(string s, uint parameterState, bool shouldReturnWhitespace)
    {
        int mark = _currentPosition;

        while (_currentPosition < _myEndOffset && !IsStringAtPosition(s) && !IsStepParameter(s))
        {
            _currentPosition++;
        }

        if (shouldReturnWhitespace)
        {
            _myState = STATE_DEFAULT;
            if (_currentPosition < _myEndOffset)
            {
                if (!IsStringAtPosition(s))
                {
                    _myState = parameterState;
                }
            }

            ReturnWhitespace(mark);
        }
    }

    private bool IsStepParameter(string currentElementTerminator)
    {
        int pos = _currentPosition;

        if (Buffer[pos] == '<')
        {
            while (pos < _myEndOffset && Buffer[pos] != '\n' && Buffer[pos] != '>' && !IsStringAtPosition(currentElementTerminator, pos))
            {
                pos++;
            }

            return pos < _myEndOffset && Buffer[pos] == '>';
        }

        return false;
    }


    public object CurrentPosition { get => _currentPosition; set => _currentPosition = (int)value; }

    public TokenNodeType TokenType { get; private set; }

    public int TokenStart => _myCurrentTokenStart;

    public int TokenEnd => _currentPosition;

    public IBuffer Buffer { get; }

    public uint State => _myState;
}